File: VialerTests/ReachabilityBarViewControllerTests.m
////
////  ReachabilityBarViewControllerTests.m
////  Copyright © 2016 VoIPGRID. All rights reserved.
////
//
//#import <XCTest/XCTest.h>
//
//#import <OCMock/OCMock.h>
//#import "ReachabilityManager.h"
//#import "ReachabilityBarViewController.h"
//#import "SystemUser.h"
//#import "TestReachabilityManagerDelegate.h"
//
//@interface ReachabilityBarViewControllerTests : XCTestCase
//@end
//
//@interface ReachabilityBarViewController ()
//@property (strong, nonatomic) ReachabilityManager *reachabilityManager;
//@property (weak, nonatomic) IBOutlet UILabel *informationLabel;
//@property (strong, nonatomic) SystemUser *currentUser;
//
//- (void)updateLayout;
//@end
//
//@implementation ReachabilityBarViewControllerTests
//
//// offline: statusbar displays: "no connection, cannot call".
//- (void)testSipNotAllowedOffline {
//    ReachabilityManagerStatusType reachabilityManagerStatus = ReachabilityManagerStatusOffline;
//    BOOL sipEnabled = NO;
//
//    BOOL statusBarVisible = YES;
//    NSString *expectedInformationLabelText = NSLocalizedString(@"No connection, cannot call.", nil);
//
//    [self performReachabilityBarUpdateLayoutTestWithReachabilityManagerStatus:reachabilityManagerStatus
//                                                     andCurrentUserSipEnabled:sipEnabled
//                                              andExpectedInformationLabelText:expectedInformationLabelText
//                                                           andShouldBeVisible:statusBarVisible];
//}
//
//// User is "allowed to SIP" but has disabled VoIP in settings:
//// 3g: statusbar displays: "VoIP disabled, enable in settings".
//- (void)testSipAllowedButDisabledLowSpeedConnection {
//    ReachabilityManagerStatusType reachabilityManagerStatus = ReachabilityManagerStatusLowSpeed;
//    BOOL sipEnabled = NO;
//
//    BOOL statusBarVisible = YES;
//    NSString *expectedInformationLabelText = NSLocalizedString(@"VoIP disabled, enable in settings", nil);
//
//    [self performReachabilityBarUpdateLayoutTestWithReachabilityManagerStatus:reachabilityManagerStatus
//                                                     andCurrentUserSipEnabled:sipEnabled
//                                              andExpectedInformationLabelText:expectedInformationLabelText
//                                                           andShouldBeVisible:statusBarVisible];
//}
//
//// 4g: statusbar displays: "VoIP disabled, enable in settings".
//// wifi: statusbar displays: "VoIP disabled, enable in settings".
//- (void)testSipAllowedButDisabledHighSpeedConnection {
//    ReachabilityManagerStatusType reachabilityManagerStatus = ReachabilityManagerStatusHighSpeed;
//    BOOL sipEnabled = NO;
//
//    BOOL statusBarVisible = YES;
//    NSString *expectedInformationLabelText = NSLocalizedString(@"VoIP disabled, enable in settings", nil);
//
//    [self performReachabilityBarUpdateLayoutTestWithReachabilityManagerStatus:reachabilityManagerStatus
//                                                     andCurrentUserSipEnabled:sipEnabled
//                                              andExpectedInformationLabelText:expectedInformationLabelText
//                                                           andShouldBeVisible:statusBarVisible];
//}
//
//// offline: statusbar displays: "no connection, cannot call".
//- (void)testSipAllowedButDisabledButOffline {
//    ReachabilityManagerStatusType reachabilityManagerStatus = ReachabilityManagerStatusOffline;
//    BOOL sipEnabled = NO;
//
//    BOOL statusBarVisible = YES;
//    NSString *expectedInformationLabelText = NSLocalizedString(@"No connection, cannot call.", nil);
//
//    [self performReachabilityBarUpdateLayoutTestWithReachabilityManagerStatus:reachabilityManagerStatus
//                                                     andCurrentUserSipEnabled:sipEnabled
//                                              andExpectedInformationLabelText:expectedInformationLabelText
//                                                           andShouldBeVisible:statusBarVisible];
//}
//
//// User is "allowed to SIP" and has enabled VoIP in settings:
//// 3g: statusbar displays: "Poor connection, Two step calling enabled".
//- (void)testSipAllowedAndEnabledLowSpeedConnection {
//    ReachabilityManagerStatusType reachabilityManagerStatus = ReachabilityManagerStatusLowSpeed;
//    BOOL sipEnabled = YES;
//
//    BOOL statusBarVisible = YES;
//    NSString *expectedInformationLabelText = NSLocalizedString(@"Poor connection, Two step calling enabled.", nil);
//
//    [self performReachabilityBarUpdateLayoutTestWithReachabilityManagerStatus:reachabilityManagerStatus
//                                                     andCurrentUserSipEnabled:sipEnabled
//                                              andExpectedInformationLabelText:expectedInformationLabelText
//                                                           andShouldBeVisible:statusBarVisible];
//}
//
//// 4g: no statusbar and no message shown.
//// wifi: no statusbar and no message shown.
//- (void)testSipAllowedAndEnabledHighSpeedConnection {
//    ReachabilityManagerStatusType reachabilityManagerStatus = ReachabilityManagerStatusHighSpeed;
//    BOOL sipEnabled = YES;
//
//    BOOL statusBarVisible = NO;
//    NSString *expectedInformationLabelText = @"";
//
//    [self performReachabilityBarUpdateLayoutTestWithReachabilityManagerStatus:reachabilityManagerStatus
//                                                     andCurrentUserSipEnabled:sipEnabled
//                                              andExpectedInformationLabelText:expectedInformationLabelText
//                                                           andShouldBeVisible:statusBarVisible];
//}
//
//// offline: statusbar displays: "no connection, cannot call.
//- (void)testSipAllowedAndEnabledButOffline {
//    ReachabilityManagerStatusType reachabilityManagerStatus = ReachabilityManagerStatusOffline;
//    BOOL sipEnabled = YES;
//
//    BOOL statusBarVisible = YES;
//    NSString *expectedInformationLabelText = NSLocalizedString(@"No connection, cannot call.", nil);
//
//    [self performReachabilityBarUpdateLayoutTestWithReachabilityManagerStatus:reachabilityManagerStatus
//                                                     andCurrentUserSipEnabled:sipEnabled
//                                              andExpectedInformationLabelText:expectedInformationLabelText
//                                                           andShouldBeVisible:statusBarVisible];
//}
//
///**
// *  A helper function for testing the updateLayout method of the ReachabilityBarViewController.
// *
// *  The following are input parameters:
// *  @param reachabilityMangerStatus The status which the ReachabilityManager should return.
// *  @param sipAllowed               Sip Allowed value for which to run the test.
// *  @param sipEnabled               Sip Enabled value for which to run the test.
// *
// *  These are the expected parameters:
// *  @param informationLabelText     The text which should appear in the information label of the bar.
// *  @param barShouldBeVisible       To test if the bar should be visible for the given input parameters
// */
//- (void)performReachabilityBarUpdateLayoutTestWithReachabilityManagerStatus:(ReachabilityManagerStatusType)reachabilityMangerStatus
//                                                   andCurrentUserSipEnabled:(BOOL)sipEnabled
//                                            andExpectedInformationLabelText:(NSString *)informationLabelText
//                                                         andShouldBeVisible:(BOOL)barShouldBeVisible {
//    // Setup
//    id mockReachabilityManager = OCMClassMock([ReachabilityManager class]);
//
//    ReachabilityBarViewController *reachabilityBarViewControllerUnderTest = [[ReachabilityBarViewController alloc] init];
//    reachabilityBarViewControllerUnderTest.reachabilityManager = mockReachabilityManager;
//
//    id mockCurrentUser = OCMClassMock([SystemUser class]);
//    reachabilityBarViewControllerUnderTest.currentUser = mockCurrentUser;
//
//    id mockReachabilityBarViewControllerDelegate = OCMStrictClassMock([TestReachabilityManagerDelegate class]);
//    reachabilityBarViewControllerUnderTest.delegate = mockReachabilityBarViewControllerDelegate;
//
//    id mockInformationLabel = OCMClassMock([UILabel class]);
//    reachabilityBarViewControllerUnderTest.informationLabel = mockInformationLabel;
//
//    // Given
//    OCMStub([mockReachabilityManager reachabilityStatus]).andReturn(reachabilityMangerStatus);
//    OCMStub([mockCurrentUser sipEnabled]).andReturn(sipEnabled);
//
//    OCMExpect([mockInformationLabel setText:informationLabelText]);
//    OCMExpect([mockReachabilityBarViewControllerDelegate reachabilityBar:reachabilityBarViewControllerUnderTest shouldBeVisible:barShouldBeVisible]);
//    OCMExpect([mockReachabilityBarViewControllerDelegate reachabilityBar:reachabilityBarViewControllerUnderTest
//                                                           statusChanged:reachabilityMangerStatus]);
//
//    // When
//    [reachabilityBarViewControllerUnderTest updateLayout];
//
//    // Then
//    OCMVerifyAllWithDelay(mockInformationLabel, 0.5);
//
//    // Cleanup
//    [mockReachabilityManager stopMocking];
//    reachabilityBarViewControllerUnderTest = nil;
//    [mockCurrentUser stopMocking];
//    [mockReachabilityBarViewControllerDelegate stopMocking];
//}
//
//
//@end

File: VialerTests/ReachabilityManager/ReachabilityBarViewControllerTests.swift
//
//  ReachabilityBasViewControllerTests.swift
//  Copyright © 2018 VoIPGRID. All rights reserved.
//

import Foundation
import XCTest

@testable import Vialer

/// TEST NEEDS FIXING WHEN THE SYSTEMUSER HAS BEEN WRITTEN IN SWIFT!
class ReachabilityBarViewControllerTests: XCTestCase {

    /// Offline: Statusbar displays: "No connection, cannot call".
    func test_SipNotAllowedOffline() {
        let status : NetworkStatus = .notReachable
        let sipEnabled: Bool = false
        let statusBarIsHidden: Bool = false
        let twoStepButtonIsHidden: Bool = true

        let expectedInformationLabelText = NSLocalizedString("No connection, cannot call.", comment: "")

        performReachabilityBarUpdateLayoutTest(reachabilityStatus: status, sipEnabled: sipEnabled, expectedInformationLabelText: expectedInformationLabelText, twoStepButtonIsHidden: twoStepButtonIsHidden, statusBarIsHidden: statusBarIsHidden)
    }

    /// User is "allowed to SIP" but has disbaled VoIP in the settings:
    /// 3G+: Statusbar displays: "VoIP disabled, enable in settings".
    func test_SipAllowedButDisabled3GPlusConnection() {
        let status: NetworkStatus = .reachableVia3GPlus
        let sipEnabled: Bool = false
        let statusBarIsHidden: Bool = false
        let twoStepButtonIsHidden: Bool = true

        let expectedInformationLabelText = NSLocalizedString("VoIP disabled, enable in settings", comment: "")

         performReachabilityBarUpdateLayoutTest(reachabilityStatus: status, sipEnabled: sipEnabled, expectedInformationLabelText: expectedInformationLabelText, twoStepButtonIsHidden: twoStepButtonIsHidden, statusBarIsHidden: statusBarIsHidden)
    }

    /// User is "allowed to SIP" but has disbaled VoIP in the settings:
    /// 4G: Statusbar displays: "VoIP disabled, enable in settings".
    func test_SipAllowedButDisabled4GConnection() {
        let status: NetworkStatus = .reachableVia4G
        let sipEnabled: Bool = false
        let statusBarIsHidden: Bool = false
        let twoStepButtonIsHidden: Bool = true

        let expectedInformationLabelText = NSLocalizedString("VoIP disabled, enable in settings", comment: "")

        performReachabilityBarUpdateLayoutTest(reachabilityStatus: status, sipEnabled: sipEnabled, expectedInformationLabelText: expectedInformationLabelText, twoStepButtonIsHidden: twoStepButtonIsHidden, statusBarIsHidden: statusBarIsHidden)
    }

    /// User is "allowed to SIP" but has disbaled VoIP in the settings:
    /// WiFi: Statusbar displays: "VoIP disabled, enable in settings".
    func test_SipAllowedButDisabledWiFiConnection() {
        let status: NetworkStatus = .reachableViaWiFi
        let sipEnabled: Bool = false
        let statusBarIsHidden: Bool = false
        let twoStepButtonIsHidden: Bool = true

        let expectedInformationLabelText = NSLocalizedString("VoIP disabled, enable in settings", comment: "")

        performReachabilityBarUpdateLayoutTest(reachabilityStatus: status, sipEnabled: sipEnabled, expectedInformationLabelText: expectedInformationLabelText, twoStepButtonIsHidden: twoStepButtonIsHidden, statusBarIsHidden: statusBarIsHidden)
    }

    func test_SipAllowedButDisabledOffline() {
        let status: NetworkStatus = .notReachable
        let sipEnabled: Bool = false
        let statusBarIsHidden: Bool = false
        let twoStepButtonIsHidden: Bool = true

        let expectedInformationLabelText = NSLocalizedString("No connection, cannot call.", comment: "")

        performReachabilityBarUpdateLayoutTest(reachabilityStatus: status, sipEnabled: sipEnabled, expectedInformationLabelText: expectedInformationLabelText, twoStepButtonIsHidden: twoStepButtonIsHidden, statusBarIsHidden: statusBarIsHidden)
    }

    /// User is "allowed to SIP" and has enabled VoIP in the settings:
    /// 3G: Statusbar displays: "Poor connection, Two step calling enabled".
    func test_SipAllowedAndEnabledBut3GConnection() {
        let status: NetworkStatus = .reachableVia3G
        let sipEnabled: Bool = true
        let statusBarIsHidden: Bool = false
        let twoStepButtonIsHidden: Bool = false

        let expectedInformationLabelText = NSLocalizedString("Poor connection, Two step calling enabled.", comment: "")

        performReachabilityBarUpdateLayoutTest(reachabilityStatus: status, sipEnabled: sipEnabled, expectedInformationLabelText: expectedInformationLabelText, twoStepButtonIsHidden: twoStepButtonIsHidden, statusBarIsHidden: statusBarIsHidden)
    }

    /// User is "allowed to SIP" and has enabled VoIP in the settings:
    /// 2G: Statusbar displays: "Poor connection, Two step calling enabled".
    func test_SipAllowedAndEnabledBut2GConnection() {
        let status: NetworkStatus = .reachableVia3G
        let sipEnabled: Bool = true
        let statusBarIsHidden: Bool = false
        let twoStepButtonIsHidden: Bool = false

        let expectedInformationLabelText = NSLocalizedString("Poor connection, Two step calling enabled.", comment: "")

        performReachabilityBarUpdateLayoutTest(reachabilityStatus: status, sipEnabled: sipEnabled, expectedInformationLabelText: expectedInformationLabelText, twoStepButtonIsHidden: twoStepButtonIsHidden, statusBarIsHidden: statusBarIsHidden)
    }

    /// User is "allowed to SIP" and has enabled VoIP in the settings:
    /// WiFi: Status bar is hidden
    func test_SipAllowedAndEnabledAndWiFiConnection() {
        let status: NetworkStatus = .reachableViaWiFi
        let sipEnabled: Bool = true
        let statusBarIsHidden: Bool = true
        let twoStepButtonIsHidden: Bool = true

        let expectedInformationLabelText = ""

        performReachabilityBarUpdateLayoutTest(reachabilityStatus: status, sipEnabled: sipEnabled, expectedInformationLabelText: expectedInformationLabelText, twoStepButtonIsHidden: twoStepButtonIsHidden, statusBarIsHidden: statusBarIsHidden)
    }

    /// User is "allowed to SIP" and has enabled VoIP in the settings:
    /// 4G: Status bar is hidden
    func test_SipAllowedAndEnabledAnd4GConnection() {
        let status: NetworkStatus = .reachableVia4G
        let sipEnabled: Bool = true
        let statusBarIsHidden: Bool = true
        let twoStepButtonIsHidden: Bool = true

        let expectedInformationLabelText = ""

        performReachabilityBarUpdateLayoutTest(reachabilityStatus: status, sipEnabled: sipEnabled, expectedInformationLabelText: expectedInformationLabelText, twoStepButtonIsHidden: twoStepButtonIsHidden, statusBarIsHidden: statusBarIsHidden)
    }

    /// User is "allowed to SIP" and has enabled VoIP in the settings:
    /// 3G+: Status bar is hidden
    func test_SipAllowedAndEnabledAnd3GPlusConnection() {
        let status: NetworkStatus = .reachableVia3GPlus
        let sipEnabled: Bool = true
        let statusBarIsHidden: Bool = true
        let twoStepButtonIsHidden: Bool = true

        let expectedInformationLabelText = ""

        performReachabilityBarUpdateLayoutTest(reachabilityStatus: status, sipEnabled: sipEnabled, expectedInformationLabelText: expectedInformationLabelText, twoStepButtonIsHidden: twoStepButtonIsHidden, statusBarIsHidden: statusBarIsHidden)
    }

    fileprivate func performReachabilityBarUpdateLayoutTest(reachabilityStatus: NetworkStatus, sipEnabled: Bool, expectedInformationLabelText: String, twoStepButtonIsHidden: Bool, statusBarIsHidden: Bool) {


        let reachabilityBarViewControllerUnderTest: ReachabilityBarViewController = ReachabilityBarViewController()

        let mockReachability = ReachabilityMock(true)!
        reachabilityBarViewControllerUnderTest.reachability = mockReachability

//        let mockCurrentUser = SystemUserMock()
//        reachabilityBarViewControllerUnderTest.currentUser = mockCurrentUser

        let mockInformationLabel = LabelMock()
        mockInformationLabel.textToShow = expectedInformationLabelText
        reachabilityBarViewControllerUnderTest.informationLabel = mockInformationLabel

        let mockTwoStepButton = ButtonMock()
        mockTwoStepButton.buttonHidden = twoStepButtonIsHidden
        reachabilityBarViewControllerUnderTest.twoStepButton = mockTwoStepButton

        /// Given
        mockReachability.statusToReturn = reachabilityStatus

        XCTAssertTrue(reachabilityBarViewControllerUnderTest.informationLabel.text!.isEqual(expectedInformationLabelText))

        reachabilityBarViewControllerUnderTest.updateLayout()

        let mockInformationLabelExists = NSPredicate(format: "isHidden == false")
        expectation(for: mockInformationLabelExists, evaluatedWith: mockInformationLabel, handler: nil)

        waitForExpectations(timeout: 5, handler: nil)
        XCTAssertTrue(reachabilityBarViewControllerUnderTest.informationLabel.isHidden == statusBarIsHidden)
        XCTAssertTrue(reachabilityBarViewControllerUnderTest.twoStepButton.isHidden == twoStepButtonIsHidden)
    }
}
